### PR TITLE
feat(RELEASE-910): add run-collectors task

### PR DESCRIPTION
**Pull request by johnbieren**
This commit adds a new task run-collectors which will run all defined collectors in the ReleasePlan and ReleasePlanAdmission and write the results to a passed resultsDir.

### CODE REVIEW

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [inline tasks/run-collectors/run-collectors.yaml]: Have you considered using an array as they suggest on the shellcheck page?
https://www.shellcheck.net/wiki/SC2086
```sh
options=(-j 5 -B)
[[ $debug == "yes" ]] && options=("${options[@]}" -d)
make "${options[@]}" file
```
(This should work even if the array ends up being empty.)

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [inline tasks/run-collectors/run-collectors.yaml]: What's this? So all the collectors will be in one repo that we'll maintain? You wanna merge it with this example value and change later? In that case at least add like a TODO comment or something?

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [inline tasks/run-collectors/run-collectors.yaml]: Isn't there a risk that two collectors will have the same name? Will we have a CI check in the git repo to prevent that?

File: tasks/run-collectors/tests/test-run-collectors-no-results-dir.yaml
Comment (mmalina) [inline tasks/run-collectors/tests/test-run-collectors-no-results-dir.yaml:1]: Can you rename this to `test-run-collectors-fail-no-results-dir` to make it immediately obvious that this is a failing scenario?

File: tasks/run-collectors/run-collectors.yaml
Comment (johnbieren) [reply tasks/run-collectors/run-collectors.yaml]: From the Jira
`This task will assume that collectors are Python scripts  defined in a managed collectors repository`
I will add a TODO

File: tasks/run-collectors/run-collectors.yaml
Comment (johnbieren) [reply tasks/run-collectors/run-collectors.yaml]: No, I don't plan to use any CI checks. There is no reason for a user to do this. If they do, they will get weird behavior, just like if a user specifies the same component name twice in their RPA mapping

File: tasks/run-collectors/run-collectors.yaml
Comment (johnbieren) [reply tasks/run-collectors/run-collectors.yaml]: I don't follow what the code you pasted does at all, and I think it is very clear how it works currently

File: tasks/run-collectors/run-collectors.yaml
Comment (johnbieren) [reply tasks/run-collectors/run-collectors.yaml]: comment added

File: tasks/run-collectors/tests/test-run-collectors-no-results-dir.yaml
Comment (johnbieren) [reply tasks/run-collectors/tests/test-run-collectors-no-results-dir.yaml:1]: updated

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [reply tasks/run-collectors/run-collectors.yaml]: ok, sounds good

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [reply tasks/run-collectors/run-collectors.yaml]: I think we should really do that, because your version would also fail if there was a space in an argument and it's not that farfetched...

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [inline tasks/run-collectors/run-collectors.yaml]:
```suggestion
                ARGS=()
                NUM_ARGS="$(jq '.params | length' <<< "$collector")"
                for ((j = 0; j < NUM_ARGS; j++)) ; do
                    param="$(jq -c --argjson j "$j" '.params[$j]' <<< "$collector")"
                    arg_name="$(jq -r '.name' <<< "$param")"
                    arg_value="$(jq -r '.value' <<< "$param")"
                    ARGS=("${ARGS[@]}" --${arg_name} "${arg_value}")
                done
                # Execute collector
                timeout "$TIMEOUT" python3 "$type" "${options[@]}" | tee "$RESULTS_DIR"/tenant-"$name".json
```

Comment (johnbieren) [reply tasks/run-collectors/run-collectors.yaml]: Awesome, thanks. I updated it and it works

File: tasks/run-collectors/run-collectors.yaml
Comment (johnbieren) [reply tasks/run-collectors/run-collectors.yaml]: fixed

**Comment by johnbieren** [comment]
@mmalina fyi, I updated this task to run in parallel since I got it figured out before this PR was merged or others reviewed it. It isn't too different, but just wanted to give you a heads up that a bit of it changed since your review

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [reply tasks/run-collectors/run-collectors.yaml]: Cool!

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [inline tasks/run-collectors/run-collectors.yaml]: -> Expected arguments are

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [inline tasks/run-collectors/run-collectors.yaml:55]: I guess we can change this later. We're probably not sure exactly what it will look like yet. But I think it would be a good idea to have a parameter for the output file in the python scripts. And it could log some stuff along the way (and of course an error if it fails) and we could save that to a file and print all the logs after the parallel runs finish - like we do here: https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/push-rpm-data-to-pyxis/push-rpm-data-to-pyxis.yaml#L171

File: tasks/run-collectors/tests/mocks.sh
Comment (mmalina) [inline tasks/run-collectors/tests/mocks.sh]: You should really use the same thing as is used in the task, so that the test is as realistic as possible: `set -xeo pipefail`

File: tasks/run-collectors/run-collectors.yaml
Comment (johnbieren) [reply tasks/run-collectors/run-collectors.yaml]: Fixed, thanks

File: tasks/run-collectors/tests/mocks.sh
Comment (johnbieren) [reply tasks/run-collectors/tests/mocks.sh]: Nice catch

File: tasks/run-collectors/run-collectors.yaml
Comment (johnbieren) [reply tasks/run-collectors/run-collectors.yaml:55]: Yeah, I don't disagree. I think that is one of maybe multiple changes to this task once we have one actual collector implemented

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [inline tasks/run-collectors/run-collectors.yaml:76]: Maybe this is not really a concern here, if all of this is reasonable fast.

But one possible improvement would be if instead of `wait -n`, you do `sleep 20s` or something. That way a slow job will not hold up the queue.

Comment (johnbieren) [reply tasks/run-collectors/run-collectors.yaml:76]: Can you explain? I thought the way this worked was if there were 4 collectors running with a concurrent limit of 4, it would wait until only 3 were running then start another

Comment (mmalina) [reply tasks/run-collectors/run-collectors.yaml:76]: AFAICS, `wait -n` will wait for the next job in the list, i.e. the first one in the list of currently running jobs. So if the one that is the oldest is slow, then even if the others are fast, we will wait for this one to finish first.

I might be wrong, but this is my understanding from reading the manual.

Comment (johnbieren) [reply tasks/run-collectors/run-collectors.yaml:76]: I think it waits for any to finish. Here is my test script
```
#!/usr/bin/env bash

set -exo pipefail

func() {
    echo "Start func call $1 at $(date)"
    sleep $1
    echo "End func call $1 at $(date)"
}

RUNNING_JOBS="\j"

for i in {10,2,3} ; do
    while (( ${RUNNING_JOBS@P} >= 2 )); do
        wait -n
    done
    func $i &
done
```
and here is the output
```
Start func call 10 at Tue Oct  1 08:44:44 PM EDT 2024
Start func call 2 at Tue Oct  1 08:44:44 PM EDT 2024
End func call 2 at Tue Oct  1 08:44:46 PM EDT 2024
Start func call 3 at Tue Oct  1 08:44:46 PM EDT 2024
End func call 3 at Tue Oct  1 08:44:49 PM EDT 2024
End func call 10 at Tue Oct  1 08:44:54 PM EDT 2024
```
func 10 is started first and takes the longest, but func 3 starts as soon as func 2 ends, not waiting for func 10 to finish

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [reply tasks/run-collectors/run-collectors.yaml:55]: ok, cool, resolving for now.

File: tasks/run-collectors/run-collectors.yaml
Comment (mmalina) [reply tasks/run-collectors/run-collectors.yaml:76]: ok, I stand corrected! Thanks for the test script! It's nice it works like that!

**Review by mmalina — APPROVED** [review]

**Pull request merged**
